src/Symfony/Component/HttpFoundation/HeaderUtils.php

```diff
@@ -193,6 +193,54 @@ public static function makeDisposition(string $disposition, string $filename, st
         return $disposition.'; '.self::toString($params, ';');
     }
 
+    /**
+     * Like parse_str(), but preserves dots in variable names.
+     */
+    public static function parseQueryString(string $qs): array
+    {
+        $query = [];
+
+        foreach (explode('&', $qs) as $v) {
+            if (false !== $i = strpos($v, "\0")) {
+                $v = substr($v, 0, $i);
+            }
+
+            if (false === $i = strpos($v, '=')) {
+                $k = urldecode($v);
+                $v = '';
+            } else {
+                $k = urldecode(substr($v, 0, $i));
+                $v = substr($v, $i);
+            }
+
+            if (false !== $i = strpos($k, "\0")) {
+                $k = substr($k, 0, $i);
+            }
+
+            if (false === $i = strpos($k, '[')) {
+                $k = bin2hex($k);
+            } else {
+                $k = substr_replace($k, bin2hex(substr($k, 0, $i)), 0, $i);
+            }
+
+            $query[] = $k.$v;
+        }
+
+        parse_str(implode('&', $query), $query);
```

Comment:
How about using this ?

```suggestion
        parse_str(http_build_query($query), $query);
```

Reply:
Not possible as this would lead to double URL encoding.